src/AspNetCore/Src/proxymodule.cxx

```diff
@@ -79,32 +78,87 @@ CProxyModule::OnExecuteRequestHandler(
-        return RQ_NOTIFICATION_FINISH_REQUEST;
+        m_pHandler = new FORWARDING_HANDLER(pHttpContext);
+        if (m_pHandler == NULL)
+        {
+            pHttpContext->GetResponse()->SetStatus(500, "Internal Server Error", 0, E_OUTOFMEMORY);
+            return RQ_NOTIFICATION_FINISH_REQUEST;
+        }
+
+        return m_pHandler->OnExecuteRequestHandler();
     }
-
-    return m_pHandler->OnExecuteRequestHandler();
+    else if (config->QueryHostingModel()->Equals(L"inprocess", true))
+    {
+        pApplicationManager = APPLICATION_MANAGER::GetInstance();
+        if (pApplicationManager == NULL)
+        {
+            hr = E_OUTOFMEMORY;
+            goto Failed;
+        }
+
+        hr = pApplicationManager->GetApplication(pHttpContext,
+            &pApplication);
+        if (FAILED(hr))
+        {
+            goto Failed;
+        }
+
+
+        hr = pApplication->GetAspNetCoreApplication(config, &pAspNetCoreApplication);
+        if (FAILED(hr))
+        {
+            goto Failed;
+        }
+
+        // Allow reading and writing to simultaneously
+        ((IHttpContext3*)pHttpContext)->EnableFullDuplex();
+
+        // Disable response buffering by default, we'll do a write behind buffering in managed code
+        ((IHttpResponse2*)pHttpContext->GetResponse())->DisableBuffering();
+
+        // TODO: Optimize sync completions
+        return pAspNetCoreApplication->ExecuteRequest(pHttpContext);
+    Failed:
+        pHttpContext->GetResponse()->SetStatus(500, "Internal Server Error", 0, E_APPLICATION_ACTIVATION_EXEC_FAILURE);
+        return REQUEST_NOTIFICATION_STATUS::RQ_NOTIFICATION_FINISH_REQUEST;
+    }
+    return REQUEST_NOTIFICATION_STATUS::RQ_NOTIFICATION_FINISH_REQUEST;
 }
 
 __override
 REQUEST_NOTIFICATION_STATUS
 CProxyModule::OnAsyncCompletion(
-    IHttpContext *,
+    IHttpContext * pHttpContext,
     DWORD                   dwNotification,
     BOOL                    fPostNotification,
     IHttpEventProvider *,
     IHttpCompletionInfo *   pCompletionInfo
 )
 {
-    UNREFERENCED_PARAMETER(dwNotification);
-    UNREFERENCED_PARAMETER(fPostNotification);
-    DBG_ASSERT(dwNotification == RQ_EXECUTE_REQUEST_HANDLER);
-    DBG_ASSERT(fPostNotification == FALSE);
+    // TODO store whether we are inproc or outofproc so we don't need to check the config everytime?
+    ASPNETCORE_CONFIG* config;
+    ASPNETCORE_CONFIG::GetConfig(pHttpContext, &config);
 
-    return m_pHandler->OnAsyncCompletion(
+    if (config->QueryHostingModel()->IsEmpty() || config->QueryHostingModel()->Equals(L"outofprocess", TRUE))// case insensitive
+    {
+        return m_pHandler->OnAsyncCompletion(
             pCompletionInfo->GetCompletionBytes(),
             pCompletionInfo->GetCompletionStatus());
+    }
+    else if (config->QueryHostingModel()->Equals(L"inprocess", TRUE)) // case insensitive
+    {
+        return REQUEST_NOTIFICATION_STATUS::RQ_NOTIFICATION_CONTINUE;
```

Comment:
do you always want to return continue even for failure case, e.g., 500 was set? Best practice is to return Finished_Request in case of error to shortcut IIS pipeline.

Reply:
@davidfowl  we probably need to add an OnAsyncCompletion handler to the AspNetCoreApplication?